hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/common/ChecksumByteBufferImpl.java

```diff
@@ -27,15 +28,31 @@ public class ChecksumByteBufferImpl implements ChecksumByteBuffer {
 
   private Checksum checksum;
 
+  private Field isReadyOnlyField = null;
+
   public ChecksumByteBufferImpl(Checksum impl) {
     this.checksum = impl;
+
+    try {
+      isReadyOnlyField = ByteBuffer.class
+          .getDeclaredField("isReadOnly");
+      isReadyOnlyField.setAccessible(true);
+    } catch (NoSuchFieldException e) {
+      e.printStackTrace();
+    }
   }
 
   @Override
   // TODO - when we eventually move to a minimum Java version >= 9 this method
   //        should be refactored to simply call checksum.update(buffer), as the
   //        Checksum interface has been enhanced to allow this since Java 9.
   public void update(ByteBuffer buffer) {
+    // this is a hack to not do memory copy.
+    try {
+      isReadyOnlyField.setBoolean(buffer, false);
```

Comment:
I must be missing something here, but isn't `buffer` defined as readonly by us here:

https://github.com/apache/ozone/blob/b3c84845926fb79e4b3b4ae20ded2be1c97ada06/hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/common/ChecksumByteBuffer.java#L42-L43

Instead of using reflection, can't we simply remove the `.asReadOnlyBuffer()` call?

Reply:
I guess there are other places that are not so simple to change, e.g. this function used by `computeChecksum` / `verifyChecksum`:

https://github.com/apache/ozone/blob/b3c84845926fb79e4b3b4ae20ded2be1c97ada06/hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/common/utils/BufferUtils.java#L68-L75

Reply:
Its the protobuf that originates the buffer as readOnly. When the protobuf bytes are received, the data goes into a protobuf ByteString, which is backed by a byteArray. ByteString has some methods to return a ByteBuffer, one of which is "asReadOnlyByteBuffer". This wraps the original byte array and returns a readonly buffer. At that stage the Java checksum interface only accepts a byte or byteArray. Even the java.util.zip.CRC32 implementation has an implementation to take a ByteBuffer, but this is what it does:

```
public void update(ByteBuffer buffer) {
        int pos = buffer.position();
        int limit = buffer.limit();
        assert (pos <= limit);
        int rem = limit - pos;
        if (rem <= 0)
            return;
        if (buffer instanceof DirectBuffer) {
            crc = updateByteBuffer(crc, ((DirectBuffer)buffer).address(), pos, rem);
        } else if (buffer.hasArray()) {
            crc = updateBytes(crc, buffer.array(), pos + buffer.arrayOffset(), rem);
        } else {
            byte[] b = new byte[rem];
            buffer.get(b);
            crc = updateBytes(crc, b, 0, b.length);
        }
        buffer.position(limit);
    }
```

buffer.hasArray() checks for readOnly and returns false if it is readonly, meaning we cannot get at the array, and it will end up copying it.